netmaster/main.go

```diff
@@ -42,147 +44,127 @@ type cliOpts struct {
+	if internalAddress == "" {
+		return nil, errors.New("netmaster interal-address is not set")
+	} else if err := netutils.ValidateBindAddress(internalAddress); err != nil {
+		return nil, err
 	}
 
-	// execute options
-	execOpts(&opts)
+	logrus.Infof("Using netmaster interal-address: %s", internalAddress)
 
-	// create master daemon
-	d := &daemon.MasterDaemon{
-		ListenURL:    opts.listenURL,
-		ControlURL:   opts.controlURL,
-		ClusterStore: opts.clusterStore,
-		ClusterMode:  opts.clusterMode,
+	// 6. validate infra type
+	infra := strings.ToLower(ctx.String("infra"))
+	switch infra {
+	case "aci", "default":
+		logrus.Infof("Using netmaster infra type: %s", infra)
+	default:
+		return nil, fmt.Errorf("Unknown netmaster infra type: %s", infra)
 	}
 
-	// initialize master daemon
-	d.Init()
+	return &daemon.MasterDaemon{
+		ListenURL:          externalAddress,
+		ControlURL:         internalAddress,
+		ClusterStoreDriver: dbConfigs.StoreDriver,
+		ClusterStoreURL:    dbConfigs.StoreURL, //TODO: support more than one url
+		ClusterMode:        netConfigs.Mode,
+		NetworkMode:        netConfigs.NetworkMode,
+		NetForwardMode:     netConfigs.ForwardMode,
+		NetInfraType:       infra,
+	}, nil
+}
 
+func startNetMaster(netmaster *daemon.MasterDaemon) {
+	// initialize master daemon
+	netmaster.Init()
 	// start monitoring services
-	d.InitServices()
-
+	netmaster.InitServices()
 	// Run daemon FSM
-	d.RunMasterFsm()
+	netmaster.RunMasterFsm()
+}
+
+func main() {
+	localIP, _ := netutils.GetDefaultAddr()
+	app := cli.NewApp()
+	app.Version = version.String()
+	app.Usage = "Contiv netmaster service"
+	netmasterFlags := []cli.Flag{
+		cli.StringFlag{
+			Name:   "infra, infra-type",
+			Value:  "default",
+			EnvVar: "CONTIV_NETMASTER_PLUGIN_HOST",
+			Usage:  "set netmaster infra tyoe, options [aci, default]",
+		},
+		cli.StringFlag{
+			Name:   "name, plugin-name",
+			Value:  "netplugin",
+			EnvVar: "CONTIV_NETMASTER_PLUGIN_HOST",
+			Usage:  "set netmaster plugin name for docker v2 plugin",
+		},
+		cli.StringFlag{
+			Name:   "external-address, listen-url",
+			Value:  fmt.Sprintf("0.0.0.0:%d", defaultPort),
+			EnvVar: "CONTIV_NETMASTER_EXTERNAL_ADDRESS",
+			Usage:  "set netmaster external address to listen on, used for general API service",
+		},
+		cli.StringFlag{
+			Name:   "interal-address, control-url",
+			Value:  fmt.Sprintf("%s:%d", localIP, defaultPort),
+			EnvVar: "CONTIV_NETMASTER_INTERNAL_ADDRESS",
+			Usage:  "set netmaster internal address to listen on, used for RPC and leader election",
+		},
+	}
+	app.Flags = utils.FlattenFlags(netmasterFlags, utils.BuildDBFlags(binName), utils.BuildNetworkFlags(binName), utils.BuildLogFlags(binName))
+	sort.Sort(cli.FlagsByName(app.Flags))
+	app.Action = func(ctx *cli.Context) error {
+		netmaster, err := initNetMaster(ctx)
+		if err != nil {
+			errmsg := err.Error()
+			logrus.Error(errmsg)
+			return cli.NewExitError(errmsg, (len(errmsg)%254 + 1))
```

Comment:
I still don't see the value in generating these exit codes based on the length of the message...

Reply:
done